Sources/Protocols/Registration/AsyncDependencyRegistering.swift

```diff
@@ -0,0 +1,94 @@
+//
+//  AsyncDependencyRegistering.swift
+//  DependencyInjection
+//
+//  Created by Róbert Oravec on 17.12.2024.
+//
+
+import Foundation
+
+/// A type that is able to register a dependency
+public protocol AsyncDependencyRegistering {
+    /// Factory closure that instantiates the required dependency
+    typealias Factory<Dependency: Sendable> = @Sendable (any AsyncDependencyResolving) async -> Dependency
+    
+    /// Factory closure that instantiates the required dependency with the given variable argument
+    typealias FactoryWithArgument<Dependency: Sendable, Argument: Sendable> = @Sendable (any AsyncDependencyResolving, Argument) async -> Dependency
+    
+    /// Register a dependency
+    ///
+    /// - Parameters:
+    ///   - type: Type of the dependency to register
+    ///   - scope: Scope of the dependency. If `.new` is used, the `factory` closure is called on each `resolve` call. If `.shared` is used, the `factory` closure is called only the first time, the instance is cached and it is returned for all subsequent `resolve` calls, i.e. it is a singleton
+    ///   - factory: Closure that is called when the dependency is being resolved
+    func register<Dependency: Sendable>(type: Dependency.Type, in scope: DependencyScope, factory: @escaping Factory<Dependency>) async
+    
+    /// Register a dependency with a variable argument
+    ///
+    /// The argument is typically a parameter in an initiliazer of the dependency that is not registered in the same resolver (i.e. container),
+    /// therefore, it needs to be passed in `resolve` call
+    ///
+    /// DISCUSSION: This registration method doesn't have any scope parameter for a reason.
+    /// The container should always return a new instance for dependencies with arguments as the behaviour for resolving shared instances with arguments is undefined.
+    /// Should the argument conform to ``Equatable`` to compare the arguments to tell whether a shared instance with a given argument was already resolved?
+    /// Shared instances are typically not dependent on variable input parameters by definition.
+    /// If you need to support this usecase, please, keep references to the variable singletons outside of the container.
```

Comment:
Interesting discussion. I agree that shared registrations with arguments doesn't make sense.

Reply:
this is old commend by @schwarja 😄 I implemented DI container in the past in previous company, it looked pretty much the same and here I just copied his solution 😄